### DISCUSSION TITLE
Tasmota relay does not switch/GPIO always "high"

### DISCUSSION THREAD

**Discussion by ionni73** [arendst/Tasmota#17769] [Q&A] [answered]
Hallo Tasmota community,

I am a newbie with a trivial question (I apologize beforehands for it).
I would like to open/close a 5v relay using a tasmotized Wemos d1 mini.
By pressing the "toggle" button, I should be able to open/close the relay

- I flashed the Wemos d1 mini with Tasmota: it works and I connect via browser
- I powered the 5v relay
- I configured the tasmota to "generic(0)" 
- I configured one of the pins (e.g. GPIO12) to be a relay
- I also configured the GPIO2 to LED in order to have a visual feedback

What happens:
- the LED works: the LED starts on and when I toggle "ON" it switches off (= it works, even if I would have prefered to have the LED on, when I toggle "ON" ;-) )
- the relay does not work: as soon as I connect the relay "signal pin" to the the "d1 GPIO", the relay switches (= the GPIO is always "high"). If I toggle to "ON", the relay does not switch. It just stays "high".

I tried to troubleshoot, but nothing helped
- I configured the module to "generic(18)" (as I do not understand the difference yet)
- I used different GPIOs
- I configured the GPIOs to relay_i (which I understand should behave "inverted") and relay b (which I still do not understand the meaning of)

I just can't find any similar behaviour in the net and in the forum. Youtube videos show that a tasmotized device with a GPIO configured as relay, switches the relay open, when the toggle button is "ON".

Again, sorry for the stupid question. I am sure the answer is trivial, but I am stuck... :-(

Jonathan

**Comment by barbudor** [comment, marked as answer]
LED: It depends how your LED is connected.
- If the Led is connected between GPIO and GND => Led
- If the led is connected between GPIO and VCC => Led_i (inverted)

RELAY:
- Without conneting the relay module to the Wemos, does the GPIO12 change state ?
- What is your relay ? Module ? Please provide schematic of the module and how the module is connected
You are lilkely not cabling or powering it correctly

**Reply by fabiorcosta** [reply]
@barbudor  I was searching fo issues with my schematic and came across your reply. I measured my relay and found it is a hard-drive.

Proved that by short-circuit GND with relay IN and it triggers. But tasmota on ESP32 won't. I tried Relay and relay_i on the configuration. It seems the GPIO is never coming to low-level (required by hard-drive relays.

as matter of testing, ESP32 is conected to USB and relay is powered from ESP32. The LED-bar is also the same as your drawing. 

Please advise

**Reply by barbudor** [reply]
I have no idea what you mean by 'hard-drive'. It is not any of the word I have used in this thread. 
Read again the whole conversation.

Without a correct schematic of your relay board, no help can be provided.

**Comment by ionni73** [comment]
Hi @barbudor ,

many thanks for the quick reply.

>     If the Led is connected between GPIO and GND => Led
>     If the led is connected between GPIO and VCC => Led_i (inverted)

I actually ment the internal LED of the D1 mini. I don't really have the possibility to interfere in the schematics.
Anyhow: after your comment I looked (better) in the Tasmota configuration and I discovered "LED_i".
I used LED_i and now it works.

That was easy :-)

> What is your relay ? Module ? Please provide schematic of the module and how the module is connected

I guess you mean I should provide some fritzing schematics, right? I am still not there yet. 
Thanks for pointing that out: I learned that I need to learn how to do this soon, if I want to ask help in a forum.
I will do this as soon as I can (will be out of town for a few weeks).

The module is a "1CH 5V Relais Modul, FR-4 PCB-Board" ( https://www.ebay.de/itm/274108493382 )

> You are lilkely not cabling or powering it correctly

I try to make myself understanable in words: I hope it's understandable.

The relay is connected 
- VCC to "+" (5v)
- GND to "-" and 
- In to the GPIO of the Wemos D1 mini

I also have the feeling that something is wrong in way I am wiring the relay...
I tried to invert VCC and GND, but then the relay does not start at all.

(BTW: The wemos is connected to the same power supply, with inbetween a step down converter to bring it 3,3V)

Thanks again for the quick reply.

Cheers

Jonathan

**Reply by barbudor** [reply]
> I guess you mean I should provide some fritzing schematics, right?

🤔
Around here, we still have paper and pencil. This works quite nicely to draw a diagram.

What is important is the how the relay module is designed
If it is a low-side driver like the diagram on the left, it should work although the way the led is connected could also cerate some problems
If it is a high-side driver like on the right, it will never work with the ESP that is a 3V3 device. It will only work with a standard 5V Arduino
It's quite easy to find out which one you have by testing the connections of the diode with VCC or GND

![image](https://user-images.githubusercontent.com/2996042/213940141-cb83860e-e6a0-4173-9c19-f5f3ee95d6d3.png)

**Reply by ionni73** [reply]
HI @barbudor 

Thanks again for the reply and apologies for the later reply of mine.

This sound interesting: I knew about Low-side and High-side (theoretically from books), but admittedly I never understood, when to use which one. Frankly speaking I still don't understand why the high-side driver transistor needs to be driven by a 5v and can't be driven by a 3,3V. I'll need to do some searching.
Assuming you are right:
- you say it's quite easy to test, but I don't understand how. how do I test it using the pins that the relay offers (VCC, GND, IN + NO, COM, NC)?
- how do I recognise a low-side from a high-side relay, when I buy it?

Thanks again

**Reply by barbudor** [reply]
> don't understand why the high-side driver transistor needs to be driven by a 5v 

In high-side the base of the transistor need to sink current to drive. As the voltage Emitter-Base is always around 0.7V, as soon as you drive the base to lower than 5V-0.7V = 4.3V, current will sink from the base and the transitor will be ON.
As the ESP is running out of 3V3, the base is always below 4.3V whether the GPIO is HI or LOW and the transistor is ALWAYS driving
Would not happen with a standard Arduino where HI signal is close to 5V

With a 3V3 relay, such high-side drive would work with an ESP (given you power the relay from another stronger 3V3 than the one provided by the tiny 3V3 regulator on the mini-D1). But modules with 3V3 relays are rare.

When you have the module in hand it's easy to test it for high drive or low drive.
Take your multimeter in Ohmeter and as you can see on the schematic above:
- A low side has 0 ohms between VCC and the pin of the diode with the bar mark
- A high side has 0 ohms between GND and the pin of hte diode opposite to the bar mark


It's almost impossible to distinguish a high-side from a low-sideon Ebay/AliExpress only from photos on those side. You need to either be able to read the reference of the transistor to find out if its PNP (high side) or NPN (low side)
You can purchase one, check it when in your hand and if not, try with another rmodel where the circuit around the diode, transistor is apparently different. When you have the right model/vendor, don't change praying for him not to sell random modules.

Models with optocoupler are generally easier to use with ESP but you may have to remove the led in series with the optocoupler because the 3V3 of the ESP is not enough to drive both the external led and the led inside the optocoupler (again for voltage reasons, not current)

You can also try to find 3V3 models such as this one : https://www.aliexpress.com/item/1005003769228120.html
On this photo, it seems that there is a PCB track between the GND pin and the pin of the diode opposite to the bar mark (=> high drive but 3V3). But it could also be my imagination....

Beware also that on those chinese side they may reference "low level trigger" and "high level trigger" which is the opposite of hte "drive"
- A high-side needs a low-level to activate the relay
- A low-side needs a high level to activate the relay

These models pretend to be configurable for high or low trigger but I never tested them myself : https://www.aliexpress.com/item/1005001903120199.html

Otherwise, just make your own module ...
I gave up using such cheap modules because they are random

**Reply by ionni73** [reply]
Hi @barbudor ,

Thanks again for the refresh. I guess i need some practice on that to gradp it completely 😉

While I was looking for the diode, which I could not find 🤣, I discovered a label on the back of the IC: "low level trigger". So I guess this root cause can now be excluded, right? Even if the failure looks as the one you described in the "high trigger" case (relay always on).


The low triggered relay should work driven by a 3,3v microcontroller, right?

Cheers 

Do you have another possible explanation?

**Reply by barbudor** [reply]
Yes, the explanation is that you haven't read carrefully the last 2 bullet-points of my previous message
low-level trigger => high-side drive => not working on ESP with 5V relay

And regarding finding the diode on your device, that should help

![image](https://user-images.githubusercontent.com/2996042/214693124-1cb20aed-9104-427b-966e-20cde16ea6ac.png)

**Reply by ionni73** [reply]
Hi @barbudor ,

Thanks again. Yes, you are right. I was so concentrated on the first part of your message that the last part did not stick in my mind. And when I found that text string I did not connected to your "warning".

Thanks for the patience.
And thanks for the picture: the diode is so small that I could not clearly identify the "bar".

I guess I can close this ticket.
Many thanks for your support: you made my life a bit easie r:-) 

Cheers from Berlin!

**Comment by ionni73** [comment]
Hi lionlggnd,

 

sorry for the late reply.

Please read through the answers that barbudor gave: very precise.

Bottom line: I bought a “high trigger” relay and now it works 😉

 

Cheers

 

 

Von: lionlggnd ***@***.***> 
Gesendet: Sonntag, 26. Februar 2023 19:24
An: arendst/Tasmota ***@***.***>
Cc: ionni73 ***@***.***>; Manual ***@***.***>
Betreff: Re: [arendst/Tasmota] Tasmota relay does not switch/GPIO always "high" (Discussion #17769)

 

i have same problem
i discovered that the relay is low trigger
can i make something to make it work with tasmota nodemcu?

**Reply by FVerhamme** [reply]
Hi there, to make your low level trigger relay work on an esp32 you need a logic level converter.
See here => https://www.youtube.com/watch?v=SWkfWWkr-tA

**Comment by fabiorcosta** [comment]
I meant high drive  - sorry.
Referring to your explanation whether the relay is low or high. And I got
the tests to confirm it is high being triggered by zero Vcc.

I also measured the D1 pin while triggering via browser and it shows
0,1Vcc. Wondering if this would be the issue here.


Fabio Costa
*email: ***@***.***
*twitter:* @fabiorcosta
*skype:* fabiorcosta





Em seg., 7 de abr. de 2025 às 02:34, Barbudor ***@***.***>
escreveu:

**Comment by lalo-uy** [comment]
If the realy is triggered by 0 V the it is called low drive.
You must configure the pin as Relai_x ( relay inverted )

**Reply by barbudor** [reply]
No
Low-side or High-side is refering to the position of the transistor regarding the load
A High-side is driven by a low signal
A Low-side is driven by a high signal

Which is what I wrote on Jan 24, 2023

![image](https://github.com/user-attachments/assets/5cfd1920-f8fe-468e-b43c-5faf13fbb426)

**Reply by fabiorcosta** [reply]
HEllo
See the schematic I am using:
<img width="722" alt="image" src="https://github.com/user-attachments/assets/9105d6e0-bfcb-444c-82d7-db195e7e812f" />
Does that help you understand?

I also confimed that the GPIO is going 0.01V when trigerred without anything connected to it. When I connect the GPIO to IN the trigger puts 0.4V to that PIN.
Also, confimed that with I connect IN to Ground it triggers the relay.

So what you guys suggest me to change ?

**Reply by barbudor** [reply]
No
I want the schematic of the relay board with proper parts identification

**Reply by fabiorcosta** [reply]
@barbudor see the schametic
![image](https://github.com/user-attachments/assets/9049fd18-36ea-4afc-936b-0ba7af5a996f)

Please let me know if this helps you help me

**Comment by fabiorcosta** [comment]
I got that. And confirmed it is triggering when I put Gpio to ground.
On the other hand GPIO12 when triggered (web) brings the voltage to 0,1V
and not to Zero. That is my issue. With 0,1V the releay is not triggered.


Em 7 de abr. de 2025, 11:46 -0300, Eduardo Spremolla <
***@***.***> escreveu:

If the realy is triggered by 0 V the it is called low drive.
You must configure the pin as Relai_x ( relay inverted )



El El lun, 7 abr. 2025 a la(s) 05:45, fabiorcosta ***@***.***>
escribió:

> I meant high drive - sorry.
> Referring to your explanation whether the relay is low or high. And I got
> the tests to confirm it is high being triggered by zero Vcc.
>
> I also measured the D1 pin while triggering via browser and it shows
> 0,1Vcc. Wondering if this would be the issue here.
>
>
> Fabio Costa
> *email: ***@***.***
> *twitter:* @fabiorcosta
> *skype:* fabiorcosta
>
>
>
>
>
> Em seg., 7 de abr. de 2025 às 02:34, Barbudor ***@***.***>
> escreveu:
>
> > I have no idea what you mean by 'hard-drive'. It is not any of the word
> I
> > have used in this thread.
> > Read again the whole conversation.
> >
> > Without a correct schematic of your relay board, no help can be
> provided.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <
>
https://github.com/arendst/Tasmota/discussions/17769#discussioncomment-12746114
>,
>
> > or unsubscribe
> > <
>
https://github.com/notifications/unsubscribe-auth/ATTUX7QD6MREJSW3L62ZKV32YIE5XAVCNFSM6AAAAAB2SGACC6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZUGYYTCNA
>
>
> > .
> > You are receiving this because you commented.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <
https://github.com/arendst/Tasmota/discussions/17769#discussioncomment-12748149
>,
> or unsubscribe
> <
https://github.com/notifications/unsubscribe-auth/ACXBW4MEGHLHAU7EIVRJ7L32YI3LXAVCNFSM6AAAAAB2SGACC6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZUHAYTIOI
>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

—
Reply to this email directly, view it on GitHub
<https://github.com/arendst/Tasmota/discussions/17769#discussioncomment-12752416>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/ATTUX7WVDZD753YIDX2BZOD2YKFTXAVCNFSM6AAAAAB2SGACC6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZVGI2DCNQ>
.
You are receiving this because you were mentioned.Message ID:
***@***.***>

**Reply by barbudor** [reply]
0.1V is a normal "low" signal for an ESP
Do not expect 0.0

If that is not enough to activate the relay, you must have another problem somewhere

Normaly with a high-drive (aka low-trigger) and an ESP device, your relay should always be activated

Without a proper and complete schematic showing ALL components (I mean including all components on the relay board), it's hard to debug remotely.

**Reply by barbudor** [reply]
Exemple schematic for a low-side relay driver which should be triggered by a high signal
This is my recommended setup for an ESP
![image](https://github.com/user-attachments/assets/cc2efc48-e3a5-4da0-b3dc-08c0f8ef0d37)

**Comment by fabiorcosta** [comment]
I will deep dive. Acquired on AliExpress.

On Wed, Apr 9, 2025 at 8:48 AM Barbudor ***@***.***> wrote:

> No
> I want the schematic of the relay board with proper parts identification
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/arendst/Tasmota/discussions/17769#discussioncomment-12777240>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ATTUX7WNGZALVGVXH7VHP6T2YUCH3AVCNFSM6AAAAAB2SGACC6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZXG4ZDIMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Fabio Costa
*email: ***@***.***
*twitter:* @fabiorcosta
*skype:* fabiorcosta

**Comment by mswiniuch** [comment]
I'm having the same issue, and I'd like to add: with the exact same configuration, the relays work fine in ESPHome. So it's definitely a problem with the Tasmota software.
